packages/cli/src/controllers/binaryData.controller.ts

```diff
@@ -0,0 +1,61 @@
+import { createReadStream } from 'node:fs';
+import { Service } from 'typedi';
+import express from 'express';
+import { BinaryDataService, FileNotFoundError, isValidNonDefaultMode } from 'n8n-core';
+import { Get, RestController } from '@/decorators';
+import { BinaryDataRequest } from '@/requests';
+
+@RestController('/binary-data')
+@Service()
+export class BinaryDataController {
+	constructor(private readonly binaryDataService: BinaryDataService) {}
+
+	@Get('/')
+	async get(req: BinaryDataRequest, res: express.Response) {
+		const { id: binaryDataId, action } = req.query;
+
+		if (!binaryDataId) {
+			return res.status(400).end('Missing binary data ID');
+		}
+
+		if (!binaryDataId.includes(':')) {
+			return res.status(400).end('Missing binary data mode');
+		}
+
+		const [mode] = binaryDataId.split(':');
+
+		if (!isValidNonDefaultMode(mode)) {
+			return res.status(400).end('Invalid binary data mode');
+		}
+
+		let { fileName, mimeType } = req.query;
+
+		try {
+			const binaryFilePath = this.binaryDataService.getPath(binaryDataId);
+
+			if (!fileName || !mimeType) {
+				try {
+					const metadata = await this.binaryDataService.getMetadata(binaryDataId);
+					fileName = metadata.fileName;
+					mimeType = metadata.mimeType;
+					res.setHeader('Content-Length', metadata.fileSize);
+				} catch {}
+			}
+
+			if (mimeType) res.setHeader('Content-Type', mimeType);
+
+			if (action === 'download') {
+				res.setHeader('Content-Disposition', `attachment; filename="${fileName}"`);
+			}
+
+			if (mode === 's3') {
+				return await this.binaryDataService.getAsStream(binaryDataId);
+			} else {
+				return createReadStream(binaryFilePath);
+			}
```

Comment:
can we move this bit to `BinaryDataService`, so that we only call `this.binaryDataService.getAsStream(binaryDataId)` here? The current code looks like it's leaking implementation details that should be abstracted away.